airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py

```diff
@@ -465,3 +539,172 @@ def test_given_record_is_record_object_when_stream_slices_then_use_record_data()
+                        Record({"id": "may_record_1", "updated_at": "2024-05-16"}, mock_slices[0]),
+                        Record({"id": "jun_record_0", "updated_at": "2024-06-15"}, mock_slices[1]),
+                        Record({"id": "jun_record_1", "updated_at": "2024-06-16"}, mock_slices[1]),
+                    ],
+                    name="first_stream",
+                ),
+                parent_key="id",
+                partition_field="partition_field",
+                parameters={},
+                config={},
+            )
+        ],
+        parameters={},
+        config={},
+    )
+
+    actual_slices = list(partition_router.stream_slices())
+    assert actual_slices == expected_slices
+
+
+def test_substream_checkpoints_after_each_parent_partition():
+    """
+    This test validates the specific behavior that when getting all parent records for a substream,
+    we are still updating state so that the parent stream's state is updated after we finish getting all
+    parent records for the parent slice (not just the substream)
+    """
+    mock_slices = [
+        StreamSlice(cursor_slice={"start_time": "2024-04-27", "end_time": "2024-05-27"}, partition={}),
+        StreamSlice(cursor_slice={"start_time": "2024-05-27", "end_time": "2024-06-27"}, partition={}),
+    ]
+
+    expected_slices = [
+        {"partition_field": "may_record_0", "parent_slice": {}},
+        {"partition_field": "may_record_1", "parent_slice": {}},
+        {"partition_field": "jun_record_0", "parent_slice": {}},
+        {"partition_field": "jun_record_1", "parent_slice": {}},
+    ]
+
+    expected_parent_state = [
+        {"start_time": "2024-04-27", "end_time": "2024-05-27"},
+        {"start_time": "2024-04-27", "end_time": "2024-05-27"},
+        {"start_time": "2024-05-27", "end_time": "2024-06-27"},
+        {"start_time": "2024-05-27", "end_time": "2024-06-27"},
+    ]
+
+    partition_router = SubstreamPartitionRouter(
+        parent_stream_configs=[
+            ParentStreamConfig(
+                stream=MockIncrementalStream(
+                    slices=mock_slices,
+                    records=[
+                        Record({"id": "may_record_0", "updated_at": "2024-05-15"}, mock_slices[0]),
+                        Record({"id": "may_record_1", "updated_at": "2024-05-16"}, mock_slices[0]),
+                        Record({"id": "jun_record_0", "updated_at": "2024-06-15"}, mock_slices[1]),
+                        Record({"id": "jun_record_1", "updated_at": "2024-06-16"}, mock_slices[1]),
+                    ],
+                    name="first_stream",
+                ),
+                incremental_dependency=True,
+                parent_key="id",
+                partition_field="partition_field",
+                parameters={},
+                config={},
+            )
+        ],
+        parameters={},
+        config={},
+    )
+
+    expected_counter = 0
+    for actual_slice in partition_router.stream_slices():
```

Comment:
nit: This seems a bit weird to me. As a user, it means that I can't call `list(partition_router.stream_slices())` else it would update the state even though the slice has not been consumed. Should we document somewhere that we assume the processing of a slice is expected to be done before the user call `next` when `incremental_dependency = True`?